Mathlib/Topology/MetricSpace/Ultra/Basic.lean

```diff
@@ -0,0 +1,234 @@
+/-
+Copyright (c) 2024 Yakov Pechersky. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yakov Pechersky
+-/
+import Mathlib.Topology.MetricSpace.Pseudo.Defs
+
+/-!
+## Ultrametric spaces
+
+This file defines ultrametric spaces, implemented as a mixin on the `Dist`,
+so that it can apply on pseudometric spaces as well.
+
+## Main definitions
+
+* `IsUltrametricDist X`: Annotates `dist : X → X → ℝ` as respecting the ultrametric inequality
+  of `dist(x, z) ≤ max {dist(x,y), dist(y,z)}`
+
+## Implementation details
+
+The mixin could have been defined as a hypothesis to be carried around, instead of relying on
+typeclass synthesis. However, since we declare a (pseudo)metric space on a type using
+typeclass arguments, one can declare the ultrametricity at the same time.
+
+TODO: Generalize to ultrametric uniformities
+
+## Tags
+
+ultrametric, nonarchimedean
+-/
+
+variable {X : Type*}
+
+/-- The `dist : X → X → ℝ` respects the ultrametric inequality
+of `dist(x, z) ≤ max (dist(x,y)) (dist(y,z))`. -/
+class IsUltrametricDist (X : Type*) [Dist X] : Prop where
```

Comment:
Have you tried to combine it with `NonarchAddGroupNorm`?

Reply:
Not in this PR, but I have a follow-up branch where I will link this to the file where `IsNonarchimedean` is used. `NonarchAddGroupNorm` isn't used anywhere outside of the file where it is defined.